src/Microsoft.Azure.WebJobs.Host/Scale/ConcurrencyManager.cs

```diff
@@ -0,0 +1,228 @@
+            if (concurrencyThrottleManager == null)
+            {
+                throw new ArgumentNullException(nameof(concurrencyThrottleManager));
+            }
+
+            _options = options;
+            _logger = loggerFactory.CreateLogger(LogCategories.Concurrency);
+            _concurrencyThrottleManager = concurrencyThrottleManager;
+
+            EffectiveCoresCount = Utility.GetEffectiveCoresCount();
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether dynamic concurrency is enabled.
+        /// </summary>
+        public bool Enabled => _options.Value.DynamicConcurrencyEnabled;
+
+        /// <summary>
+        /// Gets a value indicating whether any throttles are currently enabled.
+        /// </summary>
+        internal virtual bool ThrottleEnabled => _throttleStatus?.State == ThrottleState.Enabled;
+
+        internal int EffectiveCoresCount { get; set; }
+
+        internal ConcurrentDictionary<string, ConcurrencyStatus> ConcurrencyStatuses = new ConcurrentDictionary<string, ConcurrencyStatus>(StringComparer.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Get the concurrency status for the specified function. 
+        /// </summary>
+        /// <param name="functionId">This should be the full ID, as returned by <see cref="FunctionDescriptor.ID"/>.</param>
+        /// <returns>The updated concurrency status.</returns>
+        /// <remarks>
+        /// Update shouldn't be called concurrently for the same function ID.
+        /// </remarks>
+        public ConcurrencyStatus GetStatus(string functionId)
+        {
+            if (string.IsNullOrEmpty(functionId))
+            {
+                throw new ArgumentNullException(nameof(functionId));
+            }
+
+            // because Update won't be called concurrently for the same function ID, we can make
+            // updates to the function specific status below without locking.
+            var functionConcurrencyStatus = GetFunctionConcurrencyStatus(functionId);
+
+            if (!functionConcurrencyStatus.CanAdjustConcurrency())
+            {
+                // if we've made an adjustment recently for this function, just return the
+                // current status
+                return functionConcurrencyStatus;
+            }
+
+            // determine whether any throttles are currently enabled
+            _throttleStatus = _concurrencyThrottleManager.GetStatus();
+            if (_throttleStatus.State == ThrottleState.Unknown)
+            {
+                // if we're un an unknown state, we'll make no moves
+                // however, we will continue to take work at the current concurrency level
+                return functionConcurrencyStatus;
+            }
+
+            if (_throttleStatus.State == ThrottleState.Disabled)
+            {
+                if (CanIncreaseConcurrency(functionConcurrencyStatus))
+                {
+                    functionConcurrencyStatus.IncreaseConcurrency();
```

Comment:
The name of this method is `GetStatus` but it appears to have side-effects. I suggest giving it another name (maybe `RefreshStatus`?) so that it's more clear to consumers that this method call has side-effects.

Reply:
In my mind, from the caller's perspective they only have to know that GetStatus returns the current status. To return the current status we're making internal adjustments as necessary, butI don't think the caller needs to worry about that. E.g. we might change this in the future so that internally we've decoupled the concurrency adjustments from the GetStatus call itself (e.g. doing it on a background timer). It just happens that currently I've decided to trigger these updates when GetStatus is called.

Reply:
Looking specifically at this line of code, it seems to me that calling `GetStatus` repeatedly will drive up the concurrency numbers. That seems like a pretty important side-effect that the caller will care about since it could result in exhausting concurrency limits. Am I misunderstanding how this works?

Reply:
Internally actual concurrency changes are throttled based on the time the last increase/decrease has happened (CanAdjustConcurrency/CanIncreaseConcurrency/CanDecreaseConcurrency). Currently these are triggered based on calls to GetStatus because I like the pull model better, but we'd be at liberty in the future of changing this internally so that adjustments are made on a timer, and GetStatus doesn't trigger the increases, but just returns the current status.